### PR TITLE
Add method call info for infix operators

### LINKED ISSUE #6374: `MethodCall` info for infix operators
When a binary operator application is present in the project source module, the backend should associate a `methodPointer` and method info with the `OprApp` node. Currently it seems to do so occasionally (code making use of this information was introduced in 73487adce9040a4e4e0fcfe00704a0c9a6caeb16), but I'm not sure what circumstances reach it--usually the `methodPointer` is missing.

In particular, we need to receive the information reliably for a graph like this:
image
With method information for the `+` operator above, we will be able to display an appropriate placeholder argument similarly to how we create placeholders for non-operator method calls, fixing the secondary issue mentioned [here](https://github.com/enso-org/enso/issues/6367#issuecomment-1516435917)

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #6374

In order to provide the method pointer information, the `IrToTruffle` pass sets the module name and the type name of the builtin node.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

**Review by GregoryTravis — APPROVED** [review]

**Review by jdunkerley — APPROVED** [review]

**Review by hubertp — APPROVED** [review]
Tested the changes to builtins' registration. Seems alright and the changes are minimal.

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/IdExecutionService.java [hunk 1]
```diff
@@ -224,12 +225,12 @@ public FunctionCallInfo(FunctionCallInstrumentationNode.FunctionCall call) {
           typeName = atomConstructor.getType().getQualifiedName();
           functionName = atomConstructor.getName();
         }
-        case EnsoRootNode ensoRootNode -> {
-          moduleName = ensoRootNode.getModuleScope().getModule().getName();
-          typeName = null;
-          functionName = rootNode.getName();
+        case BuiltinRootNode builtinRootNode -> {
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/IdExecutionService.java:228]: I am surprised the case for `EnsoRootNode` is removed and nothing gets broken!?

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 2]
```diff
@@ -1692,6 +1692,187 @@ class RuntimeServerTest
     )
   }
 
+  it should "send method pointer updates of a builtin method" in {
```
Comment (JaroslavTulach) [inline engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:1695]: What is the _builtin method_? The `+` operator?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java [hunk 3]
```diff
@@ -12,6 +13,29 @@ protected BuiltinRootNode(EnsoLanguage language) {
     super(language);
   }
 
+  protected QualifiedName moduleName = null;
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java:16]: Why `protected`? There is public getter, public setter. Usually such fields are made `private` then.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java [hunk 4]
```diff
@@ -12,6 +13,29 @@ protected BuiltinRootNode(EnsoLanguage language) {
...
+  protected QualifiedName typeName = null;
+
+  /** Get the module name where the builtin is defined. */
+  public QualifiedName getModuleName() {
+    return moduleName;
+  }
+
+  /** Set the module name where the builtin is defined. */
+  public void setModuleName(QualifiedName moduleName) {
+    this.moduleName = moduleName;
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java:26]: I don't like mutability of things that should be immutable. I'd prefer to remove the setters completely. If that is not possible, please throw exception if the value is already set and somebody wants to redefine it.

File: engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala [hunk 5]
```diff
@@ -441,34 +443,47 @@ class IrToTruffle(
...
+                      new expressionProcessor.BuildFunctionBody(
+                        fn.arguments,
+                        fn.body,
+                        effectContext,
+                        true
+                      )
+                    val builtinRootNode =
+                      m.getFunction.getCallTarget.getRootNode
+                        .asInstanceOf[BuiltinRootNode]
+                    builtinRootNode.setModuleName(moduleScope.getModule.getName)
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:473]: Can't this information be passed to the factory that creates the node?

**Review by JaroslavTulach — COMMENTED** [review]
I don't like the setters!

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala (hunk 2 above)
Comment (4e6) [reply engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:1695]: Yes, `Text.+`

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/IdExecutionService.java (hunk 1 above)
Comment (4e6) [reply engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/IdExecutionService.java:228]: `EnsoRootNode` does not contain enough information to build the method pointer. We can simply skip it

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java (hunk 4 above)
Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java:26]: Ok, I'll update the setters in the next PR

File: engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala (hunk 5 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:473]: There's not enough information on that stage. The `BuiltinMethod` annotation contains an unqualified method name. There is also no information on whether the method is static or not (to derive the self-type). Although the `BuiltinType` annotation contains a qualified type name. But I'm not sure if we can correctly match builtin methods and types because there's not always a 1-1 correspondence (i.e. like with numbers subtypes)

Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:473]: CCing @hubertp as the _creator of these annotations_ and author of `Builtins::readBuiltinTypes` and `Buitins::readBuiltinMethodsMethods`.

I see that `BuiltinTypes.metadata` contains
```
Date_Time:org.enso.interpreter.node.expression.builtin.date.DateTime:Standard.Base.Data.Time.Date_Time.Date_Time:EnsoDateTime
```
which represents:
- Enso type name
- fully qualified Java class name
- fully qualified name as available in stdlib
- underlying type name of the builtin

then I also see that `BuiltinMethods.metadata` contains:
```
Date.month:org.enso.interpreter.node.expression.builtin.date.MonthDateMethodGen:false:true
```
which represents:
- the language-level name of this method
-  fully qualified Java class name
- **is static** information
- is autoregistered information

> not enough information on that stage

Given the combined content of the `BuiltinMethods.metadata` and  `BuiltinTypes.metadata` I am a believer in having all (most of) the needed information already available...

> no information on whether the method is static or not

...that information seems to be present in `BuiltinMethods.metadata`.

Should any information be missing in `BuiltinMethods.metadata` and  `BuiltinTypes.metadata` files, then I'd rather add such information there than trying to supply it later via setters in `IrToTruffle` stage. Immutability creates better systems that are easier to reason about. Having [half-initialized objects floating around the system](http://wiki.apidesign.org/wiki/Singleton#Singletons_are_Pathological_Liars.5B3.5D) is just asking for troubles.

Please accept my apology for not providing this review sooner than this PR got merged.

Comment (hubertp) [reply engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:473]: I'm not sure if this will cover static wrappers. But yes, in general that information is ready in the metadata files that are being read in static initializers.

Comment (Akirathan) [reply engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:473]: So if I understand that correctly, instead of providing the moduleName and typeName here via setters on `BuiltinRootNode`, you are suggesting to tweak [Builtins$LoadedBuiltinMethod::toFunction](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java#L650) method (and any other affected locations in the code) such that it will create a `BuiltinRootNode` with already initialized moduleName and typeName? That sounds like a good idea.

> Immutability creates better systems that are easier to reason about.

Indeed. Let me also add a note that setters are twice as much problematic in the Truffle environment than in pure Java. What if `BuiltinRootNode.setModuleName` is used during runtime? Should we manually invalidate any assumptions? Is it safe not to call `CompilerDirectives.transferToInterpreterAndInvalidate()` in these setters? Should the `moduleName` and `typeName` fields be marked as `@CompilationFinal`? (They probably should, theoretically, `BuiltinRootNode` can be shared among multiple contexts and can be a compilation constant; therefore, these fields can also be compilation constants). etc ...

TL;DR; I agree with Jaroslav, and I favor a follow-up PR that removes these setters (or tracking it in a separate issue). The sooner we do it, the better.
